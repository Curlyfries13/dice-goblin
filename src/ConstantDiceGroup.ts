import { DiceTerm } from 'DiceTerm';
import Constant from 'Constant';
import { StatisticalGenerator } from 'StatisticalGenerator';

/**
 * A "dice" group that returns a constant result
 *
 * Mostly useful for testing
 */

export default class ConstantDiceGroup implements DiceTerm {
  sides: StatisticalGenerator;

  currentSides: number;

  count: StatisticalGenerator;

  currentCount: number;

  statProps: {
    min: number;
    max: number;
    average: number;
    periodicity: number;
  };

  combinatoricMagnitude: number;

  value: () => number;

  pdf: (value: number) => number;

  multinomial: (value: number) => number;

  current: number[];

  results: number[];

  constructor(results: number[], sides?: number) {
    this.results = results;
    if (sides === undefined) {
      const defaultSides = Math.max(...results);
      this.sides = new Constant(defaultSides);
      this.currentSides = defaultSides;
    } else {
      this.sides = new Constant(sides);
      this.currentSides = sides;
    }
    const sum = results.reduce((acc, curr) => acc + curr, 0);

    this.count = new Constant(results.length);
    this.currentCount = results.length;
    this.statProps = {
      min: sum,
      max: sum,
      average: sum,
      periodicity: sum,
    };
    this.combinatoricMagnitude = 1;

    this.current = [sum].concat(this.results);
    this.value = this.roll;
    this.pdf = (value: number) => {
      if (value === this.current[0]) {
        return 1;
      }
      return 0;
    };
    this.multinomial = (value: number) => {
      if (value === this.current[0]) {
        return 1;
      }
      return 0;
    };
  }

  roll() {
    return this.rollGroup()[0];
  }

  rollGroup() {
    return this.current;
  }
}
