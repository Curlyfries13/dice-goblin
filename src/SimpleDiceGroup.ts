import random, { RNG } from 'random';
import seedrandom, { PRNG } from 'seedrandom';

import { BATCH_SIZE } from 'EngineConfig';
import Constant from 'Constant';
import { DiceTerm } from 'DiceTerm';
import { StatisticalGenerator } from 'StatisticalGenerator';
import { multinomialCoefficient, multinomialPDF } from 'utils';

/**
 * A dice group is a single type of Dice, e.g. a six-sided dice
 *
 * This Die doesn't handle modifiers, instead that should be handled by a
 * "Roll"
 *
 * The die also doesn't handle polymorphic rolls, i.e. if the number of dice or
 * the faces on a die change
 */
export default class SimpleDiceGroup implements DiceTerm, StatisticalGenerator {
  sides: StatisticalGenerator;

  currentSides: number;

  count: StatisticalGenerator;

  currentCount: number;

  statProps: {
    min: number;
    max: number;
    periodicity: number;
    average: number;
  };

  combinatoricMagnitude: number;

  seed: undefined | seedrandom;

  current: number[];

  // placeholder
  value: () => number;

  pdf: (value: number) => number;

  multinomial: (value: number) => number;

  constructor(sides: number = 6, count: number = 1, seed?: PRNG) {
    this.sides = new Constant(sides);
    this.currentSides = sides;
    this.count = new Constant(count);
    this.currentCount = count;
    this.combinatoricMagnitude = sides ** count;
    this.statProps = {
      // this die engine always uses a minimum value of 1
      // Note this acts strangely if the number of sides is 0. In this case an
      // error should be thrown
      min: count,
      max: count * sides,
      // this is a helper for PDF calculation and convolution
      periodicity: count * sides - count,
      average: ((1 + sides) / 2.0) * count,
    };

    if (seed !== undefined) random.use(seed as unknown as RNG);
    this.current = [count].concat(Array(count).fill(1));
    this.value = this.roll;
    this.pdf = (value: number) => multinomialPDF(this.count.value(), this.sides.value(), value);
    this.multinomial = (value: number) =>
      multinomialCoefficient(this.count.value(), this.sides.value(), value);
  }

  // Roll this group and return the result
  roll(): number {
    if (this.sides === undefined || this.count === undefined) {
      // the group is not defined, initialize first
      // TODO: write more descriptive errors
      // TODO: create custom error types
      throw Error('Dice Group is not configured');
    }
    return this.rollGroup()[0];
  }

  /*
   * Roll the dice group and return the results
   * The first element of the group is the sum, the following results are the
   * individual die faces
   */
  rollGroup(): number[] {
    const count = this.count.value();
    const sides = this.sides.value();
    let results: number[] = [];
    // This batch size is picked arbitrarily, but it should be below the call
    // limit for your target browsers
    let resultCount = 0;
    // avoid the call stack error by running in batches
    for (let i = 0; i < count; i += BATCH_SIZE) {
      // determine if we run a full batch or a smaller batch
      if (resultCount + BATCH_SIZE < count) {
        results = results.concat(Array.apply(0, Array(BATCH_SIZE)).map(() => random.int(1, sides)));
        resultCount += BATCH_SIZE;
      } else {
        const subBatchSize = count - resultCount;
        results = results.concat(
          Array.apply(0, Array(subBatchSize)).map(() => random.int(1, sides)),
        );
      }
    }
    const reducer = (total: number, value: number): number => total + value;
    const sum = results.reduce(reducer, 0);
    results.unshift(sum);
    this.current = results;
    return results;
  }
}
