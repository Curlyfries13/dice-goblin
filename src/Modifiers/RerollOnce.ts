import CompareMode from 'CompareMode';
import Constant from 'Constant';
import SimpleDiceGroup from 'SimpleDiceGroup';
import { DiceTerm } from 'DiceTerm';
import { Modifier } from 'Modifier';
import { StatisticalGenerator } from 'StatisticalGenerator';
import { convolution } from 'utils';

/*
 * The Reroll Once Modifier adds additional logic to the dice. Essentially, they
 * retry if any dice roll some value or lower / higher, and take the next value.
 */

export default class RerollOnce implements Modifier, DiceTerm {
  name = 'RerollOnce';

  target: StatisticalGenerator;

  compareMode: CompareMode;

  base: DiceTerm;

  baseResults: number[];

  sides: StatisticalGenerator;

  currentSides: number;

  count: StatisticalGenerator;

  currentCount: number;

  statProps: {
    min: number;
    max: number;
    average: number;
    periodicity: number;
  };

  combinatoricMagnitude: number;

  current: number[];

  reroller: StatisticalGenerator;

  value: () => number;

  pdf: (value: number) => number;

  multinomial: (value: number) => number;

  rerollCount: number;

  rerollRange: number;

  constructor(base: DiceTerm, target?: StatisticalGenerator, compare?: CompareMode) {
    this.base = base;
    // by default, a reroll happens on the minimum roll only (i.e. brutal 1's)
    // NOTE: I can't think of a way where we don't just reroll 1's
    this.target = target === undefined ? new Constant(1) : target;
    // NOTE: this _could_ be unintuitive? May be expecting LessThan?
    this.compareMode = compare === undefined ? CompareMode.Equal : compare;
    this.baseResults = [];
    this.current = [];
    this.sides = new Constant(base.sides.statProps.max);
    this.currentSides = this.sides.statProps.max;
    // Do not copy the base values: this causes weirdness when you try to modify their properties
    this.count = {
      value: this.base.count.value,
      // these properties are almost certainly not helpful
      combinatoricMagnitude: this.base.count.combinatoricMagnitude,
      statProps: {
        min: this.base.statProps.min,
        max: this.base.statProps.max,
        average: this.base.count.statProps.average,
        periodicity: this.base.sides.statProps.max,
      },
      pdf: (value: number) => this.base.count.pdf(value),
      multinomial: (value: number) => this.base.count.multinomial(value),
    };
    this.currentCount = this.count.statProps.max;

    // TODO: consolidate naming, this is similar for reroll. Maybe need to make
    // this consistent across all Statistical Generators
    this.rerollRange = 1;
    let averageMod = 0;
    const numerator = this.sides.statProps.average - this.target.statProps.average;
    const denominator = this.sides.statProps.average;
    if (compare === CompareMode.LessThan) {
      // NOTE: this assumes that we're using dice that start at one always
      this.rerollRange = this.target.statProps.average;
      averageMod = numerator <= 0 ? 0 : numerator / denominator;
    } else if (compare === CompareMode.GreaterThan) {
      // due to greaterThan actually being greater than or equal to the target
      this.rerollRange = this.sides.statProps.average - this.target.statProps.average + 1;
      averageMod = numerator <= 0 ? 0 : -1 * (numerator / denominator);
    }

    this.combinatoricMagnitude =
      this.sides.statProps.average ** (this.count.statProps.average + this.rerollRange);
    this.statProps = {
      min: base.statProps.min,
      max: base.statProps.max,
      average: base.statProps.average + averageMod,
      periodicity: base.statProps.periodicity,
    };
    this.rerollCount = 0;
    this.reroller = new SimpleDiceGroup(this.base.sides.statProps.max, 1);
    this.value = this.roll;
    // PERFORMANCE: consider memoizing convolution-heavy tests
    // TODO: get the value of a dice roll targeting a specific number (average)
    this.pdf = (value: number) => {
      // NOTE: performance consider if this is a performance hit
      if (this.base.pdf(value) === 0) return 0;
      // this is busted for counts higher than 1
      if (
        this.base.count.statProps.min === 1 &&
        this.base.count.statProps.max === 1 &&
        this.base.sides.statProps.min === this.base.sides.statProps.max
      ) {
        if (this.thresholdFunc(value)) {
          return (1 / this.rerollRange) * this.base.pdf(value);
        }
        return this.base.pdf(value) + (1 / this.rerollRange) * this.base.pdf(value);
      }
      if (
        this.base.count.statProps.min === this.base.count.statProps.max &&
        this.base.sides.statProps.min === this.base.sides.statProps.max
      ) {
        // create a convolution sub-entity
        // NOTE: performance (this could get expensive)
        const left = new RerollOnce(
          new SimpleDiceGroup(this.base.sides.value(), 1),
          this.target,
          this.compareMode,
        );
        const right = new RerollOnce(
          new SimpleDiceGroup(this.base.sides.value(), this.base.count.value() - 1),
          this.target,
          this.compareMode,
        );
        return convolution(value, left, right, (x, y) => x - y, 'pdf');
      }
      // give up
      // TODO: implement PDF for non-constant dice
      return this.base.pdf(value);
    };
    // TODO: implement this
    this.multinomial = (value: number) => {
      // NOTE: performance consider if this is a performance hit
      if (this.base.multinomial(value) === 0) return 0;
      // this is busted for counts higher than 1
      // if we only have 1 die
      if (
        this.base.count.statProps.min === 1 &&
        this.base.count.statProps.max === 1 &&
        this.base.sides.statProps.min === this.base.sides.statProps.max
      ) {
        if (this.thresholdFunc(value)) {
          return (1 / this.rerollRange) * this.base.pdf(value);
        }
        return this.base.pdf(value) + (1 / this.rerollRange) * this.base.pdf(value);
      }
      if (
        this.base.count.statProps.min === this.base.count.statProps.max &&
        this.base.sides.statProps.min === this.base.sides.statProps.max
      ) {
        // create a convolution sub-entity
        // NOTE: performance (this could get expensive)
        const left = new RerollOnce(
          new SimpleDiceGroup(this.base.sides.value(), 1),
          this.target,
          this.compareMode,
        );
        const right = new RerollOnce(
          new SimpleDiceGroup(this.base.sides.value(), this.base.count.value() - 1),
          this.target,
          this.compareMode,
        );
        return convolution(value, left, right, (x, y) => x - y, 'pdf');
      }
      // give up
      // TODO: implement PDF for non-constant dice
      return this.base.pdf(value);
    };
  }

  thresholdFunc = (value: number): boolean => {
    // TODO: fix multiple 'r' rerolls, this is apparently a thing in the spec
    switch (this.compareMode) {
      case CompareMode.LessThan:
        // according to the Roll20 spec, '<' actually means '<='
        return value <= this.target.value();
      case CompareMode.GreaterThan:
        return value >= this.target.value();
      case CompareMode.Equal:
      default:
        return value === this.target.value();
    }
  };

  roll(): number {
    return this.rollGroup()[0];
  }

  // TODO: determine if it's actually worth it to literally reroll, or if it'd
  // be better to just construct the kind of dice that would be created by
  // restricting the randomized values
  rollGroup(): number[] {
    // roll the base, if any results don't pass the threshold, roll those again.
    this.base.rollGroup();
    this.reroller = new SimpleDiceGroup(1, this.base.currentSides);
    this.rerollCount = 0;
    this.baseResults = this.base.current.slice(0);
    const active = this.baseResults.slice(1).reduce((acc, value, i) => {
      if (this.thresholdFunc(value)) {
        acc.push(i);
      }
      return acc;
    }, [] as number[]);

    for (let i = 0; i < active.length; i += 1) {
      const newValue = this.reroller.value();
      // console.log(`replace ${i} value: ${this.baseResults[active[i] + 1]} => ${newValue}`);
      this.rerollCount += 1;
      this.baseResults[active[i] + 1] = newValue;
    }
    let sum = 0;
    for (let i = 1; i < this.baseResults.length; i += 1) {
      sum += this.baseResults[i];
    }
    this.baseResults[0] = sum;
    // console.log(`post - mod: ${JSON.stringify(this.baseResults)} `);
    return this.baseResults;
  }
}
