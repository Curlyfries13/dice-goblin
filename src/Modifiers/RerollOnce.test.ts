import seedrandom from 'seedrandom';

import SimpleDiceGroup from 'SimpleDiceGroup';
import ConstantDiceGroup from 'ConstantDiceGroup';
import RerollOnce from './RerollOnce';

it('Rerolls the lowest roll by default, keeping dice count', () => {
  // this produces a roll with 2 max rolls
  // cspell:disable-next-line
  const rng = seedrandom('testrng');
  const dice = new SimpleDiceGroup(4, 10, rng);
  const rerollDice = new RerollOnce(dice);
  rerollDice.rollGroup();
  expect(rerollDice.count.value()).toMatchSnapshot();
});

it('Rerolls the lowest values by default', () => {
  const constantDice = new ConstantDiceGroup([1, 1, 1, 1], 4);
  const rerollDice = new RerollOnce(constantDice);
  rerollDice.rollGroup();
  expect(rerollDice.count.value()).toBe(4);
  expect(rerollDice.rerollCount).toBeGreaterThanOrEqual(4);
});
