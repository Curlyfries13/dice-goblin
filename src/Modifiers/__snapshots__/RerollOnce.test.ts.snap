// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rerolls the lowest roll by default, keeping dice count 1`] = `10`;
